lib/rust/ensogl/core/src/display/symbol/gpu/registry.rs

```diff
@@ -76,49 +93,42 @@ impl SymbolRegistry {
             variables,
             context,
             stats,
+            next_id,
         }
     }
 
-    /// Creates a new `Symbol` instance and returns its id.
-    pub fn new_get_id(&self) -> SymbolId {
+    /// Creates a new `Symbol`.
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(&self) -> Symbol {
         let symbol_dirty = self.symbol_dirty.clone();
         let stats = &self.stats;
-        let index = self.symbols.borrow_mut().insert_with_ix_(|ix| {
-            let id = SymbolId::new(ix as u32);
-            let on_mut = move || symbol_dirty.set(id);
-            let symbol = Symbol::new(stats, id, &self.global_id_provider, on_mut);
-            symbol.set_context(self.context.borrow().as_ref());
-            symbol
-        });
-        SymbolId::new(index as u32)
+        let id_value = self.next_id.get();
+        self.next_id.set(id_value + 1);
+        let id = SymbolId::new(id_value);
+        let on_mut = move || symbol_dirty.set(id);
+        let symbol = Symbol::new(stats, id, &self.global_id_provider, on_mut);
+        symbol.set_context(self.context.borrow().as_ref());
+        self.symbols.borrow_mut().insert(id, symbol.clone_ref());
+        symbol
     }
 
     /// Set the GPU context. In most cases, this happens during app initialization or during context
     /// restoration, after the context was lost. See the docs of [`Context`] to learn more.
     pub fn set_context(&self, context: Option<&Context>) {
         *self.context.borrow_mut() = context.cloned();
-        for symbol in &*self.symbols.borrow() {
+        for symbol in self.symbols.borrow().values() {
             symbol.set_context(context)
         }
     }
 
-    /// Creates a new `Symbol` instance.
-    #[allow(clippy::new_ret_no_self)]
-    pub fn new(&self) -> Symbol {
-        let ix = self.new_get_id();
-        self.index(ix)
-    }
-
-    /// Get symbol by its ID.
-    pub fn index(&self, id: SymbolId) -> Symbol {
-        self.symbols.borrow()[(*id) as usize].clone_ref()
-    }
-
     /// Check dirty flags and update the state accordingly.
     pub fn update(&self) {
         debug!(self.logger, "Updating.", || {
+            let symbols = self.symbols.borrow();
             for id in self.symbol_dirty.take().iter() {
-                self.symbols.borrow()[(**id) as usize].update(&self.variables)
+                if let Some(symbol) = symbols.get(id) {
+                    symbol.update(&self.variables);
+                }
```

Comment:
I know this is a minor optimization, but I don't like that we are right now querying hashmap for every symbol rendering. Earlier we've been just indexing array, which is faster. This is the main rendering loop, and every frame , there could be a lot of symbols on a stage, even 100 000. In such a case, indexing can be a lot faster than querying hashmap. We want to opimize for frame rendering speed. Right now we are OK with 60 FPS, which gives 16ms per frame. We need to reserve 10ms for engine. If we want to feel native on new macs, we should get 120 FPS in the future. This would give us 8ms, where at least 6ms should be for enigne. So basically every ms counts here.

Reply:
If I understand correctly, the number of lookups we perform here scales with the number of `Symbol`s, not with the number of `SymbolInstance`s; my measurements are consistent with this (scene with 140 symbols, 3570 sprites: 134 ids passed to `render_by_ids` calls per frame). Am I misunderstanding the performance characteristics here?

If we have 500 `Symbol`s (which would reflect a very complex scene--currently, a large portion of our `Symbol`s are because of the inefficient text area implementation), and these lookups scale with `Symbol` count and not `SymbolInstances`, then: To perform 500 lookups in a 500 element hashmap takes about 4μs on my machine; that would be 0.2% of a 2ms budget, which might be more than negligible, but is not huge.

---

There are several concerns an unregistration implementation has to balance:
- Performance, of course.
- If we reuse `SymbolId`s, we must ensure with high confidence that IDs never outlive their `Symbol`s; getting our signals crossed is always a bad failure mode.
- We currently rely on sequential assignment of `SymbolId`s to keep track of `Symbol` creation order, which we use in render ordering.

One solution I considered would allow direct-indexing without the hazards of naïve reuse, while still remembering `Symbol` creation order: We could pack into a 64-bit `SymbolId` a 48-bit globally-unique sequence number, along with a 16-bit index number. We use an `OptVec`, but along with each `WeakShader` we store the corresponding sequence number. To perform a lookup, we index by the index number, and then check that the sequence number matches the stored value. To compare `Symbol` creation order, we can use the sequence number component of the ID. I haven't implemented it this way because, if my understanding of the scaling is accurate, I don't think the performance benefit is worth the introduction of a custom data structure here.

Reply:
Thanks for a great write-up, as always. TBH, I'm way less sensitive than you regarding the point `If we reuse SymbolIds, we must ensure with high confidence that IDs never outlive their Symbols;`. I understand why you want it though. I was just always ok with the new symbol getting ID of an already destroyed one. Don't get me wrong, I highly appreciate your attitude here and I feel it's better than mine. I'm just explaining why you don't see that in the existing code.

Regarding symbols vs instances – you are right. However, when designing EnsoGL I also have different use cases in my head. Some involve a lot of symbols and not many instances. It's not our use case though.

Anyway, I still feel that `performance >>> debugging info`, and thus I'm more OK with reusing the symbols IDs rather than losing bits of performance here and there. I might be over-sensitive regarding rendering loop performance, but this is needed - otherwise, we would not work well on slower computers (your 0.2% budget can be 1% of budget there, which is already significant only for iteration of symbols for rendering).

We have 2 possible ways to go here:
1. Leave it as is, accept this minor perf hit, and link to this discussion from docs there.
2. Change it to just vector indexing, accept the fact that IDs are reused in a naive fashion, and leave there a comment to fix it in the future, again linking to this discussion.

I like the solution 2 way more than 1 TBH, as I think we should be extremely performance-oriented in our rendering engine.

Reply:
Implemented the higher-performance solution I proposed this morning.